src/PIL/DdsImagePlugin.py

```diff
@@ -148,15 +147,38 @@ def _open(self):
             self.tile = [("raw", (0, 0) + self.size, 0, (self.mode, 0, 1))]
         elif pfflags & DDPF_RGB:
             # Texture contains uncompressed RGB data
-            masks = {mask: ["R", "G", "B", "A"][i] for i, mask in enumerate(masks)}
-            rawmode = ""
-            if pfflags & DDPF_ALPHAPIXELS:
-                rawmode += masks[0xFF000000]
-            else:
+            if not pfflags & DDPF_ALPHAPIXELS:
                 self._mode = "RGB"
-            rawmode += masks[0xFF0000] + masks[0xFF00] + masks[0xFF]
 
-            self.tile = [("raw", (0, 0) + self.size, 0, (rawmode[::-1], 0, 1))]
+            # Each channel is derived using a mask of the pixel values
+            masks = {
+                mask: "RGB"[i]
+                for i, mask in enumerate(struct.unpack("<3I", header.read(12)))
+            }
+            if self.mode == "RGB" and all(
+                mask in (0b0000000000011111, 0b0000001111100000, 0b0111110000000000)
+                for mask in masks
+            ):
+                # Each mask uses 5 bits
+                rawmode = (
+                    masks[0b0000000000011111]
+                    + masks[0b0000001111100000]
+                    + masks[0b0111110000000000]
+                    + ";15"
+                )
+            else:
+                # Try 8 bits for each mask
+                if self.mode == "RGBA":
+                    # The fourth mask, alpha,
+                    # is only valid when the correct flags are present
```

Comment:
What are the correct flags?

Reply:
https://learn.microsoft.com/en-us/windows/win32/direct3ddds/dds-pixelformat seems inconsistent on this to me

> DDPF_ALPHAPIXELS | Texture contains alpha data; dwRGBAlphaBitMask contains valid data.

> Alpha mask for reading alpha data. dwFlags must include DDPF_ALPHAPIXELS or DDPF_ALPHA

The code is checking for `DDPF_ALPHAPIXELS`
https://github.com/python-pillow/Pillow/blob/aa03c5a82a0213260517ea4d1ed8925babea42ca/src/PIL/DdsImagePlugin.py#L150